### PR TITLE
Dependencies inverted for DTO's

### PR DESCRIPTION
**Pull request by ChuckMoe**
# Draft
## Description

This is an example for issue #941.
Schemas are built up from the smallest subset, in order to only allow the user to input the allowed variables.

## Motivation

Take a look at #941

## Fixes:

* Restricts API to only use (create, update, read) allowed values

## Changes:

* Instead of extending from the base class and saying only part of the data can be passed in, I reversed the dependencies and built up the variable space

## Tests included/Docs Updated?

- [ ] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 
- [ ] Docs updated?
- [ ] New packages used/requires npm install? 
- [ ] Toggle added for new features?

### CODE REVIEW

**Review by nitrosx — COMMENTED** [review]
@ChuckMoe am I correct if if I say that you define the update dto for a specific entity and from there you inherit the class for the create dto adding the missing fields?
I was wondering if you can explain why you decide to take this approach?
Thanks

**Comment by ChuckMoe** [comment]
@nitrosx
Creating the UpdateDto class and using it as a base for the CreateDto class has the advantage that you can explicitly "exclude" certain parameters you do not want to be updatable, like an id.

Here an example.
```
class UpdateDto {
  readonly color: string;
}

class CreateDto extends UpdateDto {
  readonly id: int;
}
```

Now the following would be valid requests.
```
POST /create
{
  "id": 1,
  "color": "yellow"
}

POST /update/1
{
  "color": "green"
}
```

These would be invalid requests now though.
```
POST /update/1
{
  "id": 2,
  "color": "green"
}
```

Before this PR, the UpdateDto extended the CreateDto and made it partial, which theoretically allowed for updating parameters that were not supposed to be updated as well. This is also reflected in the generated OpenAPI schema in /explorer and helps for understanding and automatic generation of code as well.

**Comment by ChuckMoe** [comment]
I would very much like to go through all of the classes with someone to identify if all of the variables are correctly placed, or if I need to move them between the update and create class.

**Comment by nitrosx** [comment]
@ChuckMoe I will be available to go over with you as a first pass. 
I will try to open start a discussion about this PR in today's meeting.

**Comment by ChuckMoe** [comment]
> I will try to open start a discussion about this PR in today's meeting.

Good morning. Any news on this @nitrosx ?

**Comment by nitrosx** [comment]
Unfortunately, I was not able to initiate the discussion at the last meeting.
Should we raise the priority?

**Comment by nitrosx** [comment]
@bpedersen2 Can you provide your opinion on this?

**Comment by bpedersen2** [comment]
Sounds like a valid idea. Lets try to put this on the agenda for tuesday (This time I will be there again).

@ChuckMoe: could you promote this from draft to an active pull request?

**Comment by nitrosx** [comment]
@ChuckMoe and @bpedersen2 I started to review once more this PR, but I realized that I quickly got lost.
I have a feeling that I/we need a well defined method to verify if we have the correct fields in the correct model.

Would a table with the fields and a column for each DTO work? Thoughts and feedback?

That said I embrace and support the idea to invert the dependencies.

**Comment by ChuckMoe** [comment]
Good morning @nitrosx. A table with a column `updatable` should be enough. That would be really awesome to have. Going through all of the DTOs will be very quick then.

**Comment by nitrosx** [comment]
@ChuckMoe can you provide an example please?
I will try to merge your PR as soon as I am able to

**Comment by ChuckMoe** [comment]
Here is an example table that would help going through all the DTO's. The _Optional_ column is not necessarily needed for this PR but good for overview and to check if everything is as it is supposed to be.

### Sample

| Variable | Optional | Can Update |
| --- | --- | --- |
| ownerGroup | False | True |
| accessGroups | True | True |
| instrumentGroups | True | True |
| sampleId | True | False |
| owner | True | True |
| description | True | True |
| sampleCharacteristics | True | True |
| isPublished | True | True |

File: src/proposals/dto/update-proposal.dto.ts [hunk 1]
```diff
@@ -1,10 +1,119 @@
...
+    type: Date,
+    required: false,
+    description: "The date when data collection finishes.",
+  })
+  @IsOptional()
+  @IsDateString()
+  readonly endTime?: Date;
+
+  @ApiProperty({
+    type: MeasurementPeriodClass,
```
Comment (fpotier) [inline src/proposals/dto/update-proposal.dto.ts]: I think this should be a `CreateMeasurementPeriodDto` rather than a `MeasurementPeriodClass`

Comment (ChuckMoe) [reply src/proposals/dto/update-proposal.dto.ts]: You are correct, thank you very much!

**Review by bpedersen2 — APPROVED** [review]

**Pull request merged**